### ISSUE TITLE
Flakiness when using the EVA protocol

### ISSUE THREAD

**Issue by rmadhwal** [Tribler/trustchain-superapp#101]
Hi, we're using the EVA protocol to download apps as a fallback in Freedom of Computing, while experimenting with the protocol I noticed that transfers would time out about half the time.

On investigation, I found that this was due to data packets not making it through to the requester and this resulting in multiple reacknowledgements which inevitably lead to the Timeout Errors. I tried increased the retransmit interval and retransmit attempt counts (along with the overall timeout), but this didn't yield in much success either. Finally, I was able to obtain some consistency with a slightly increased timeout and a smaller block size (1200 -> 600), I suspect that larger packets are somehow getting dropped over UDP and perhaps the right way forward is worth looking more into.

**Comment by drew2a** [comment]
We see quite similar behavior of EVA protocol that was revealed during experiments made by @devos50.
I fixed all the problems found in this PR: https://github.com/Tribler/tribler/pull/6831

Hope it will help you as well.

**Comment by synctext** [comment]
excellent progress. At the final day of week 8 it is coming together. App now installs and actually provides freedom :boom: 

You now provide ```rough``` ability to execute any code and yield control to your personal devices, lessen the power of Big Tech.

**Comment by synctext** [comment]
thnx for that reference! That is great info to see, attempting only a single write to sockets is quite a bug in my world. Very much needs fixing if Kotlin code has this bug also :construction:

**Comment by drew2a** [comment]
Regarding the only one attempt to send `WriteRequest`. I should say that it was done intentionally.

My arguments were:

1. Message retransmitting by itself adds complexity to the protocol and could lead to hidden bugs.
2. An user of EVA will handle failures of sending `WriteRequest` on the application layer. Why? Because `WriteRequest` is the very first packet sent to the peer and it can be rejected/dropped for multiple reasons. These failure handlers could differ from application to application.
3. I assumed that these handlers will be presented anyway. That's why it is a reason to move the responsibility of handling `WriteRequest` failures to the higher layer (e.g. application).

But now we have three examples of using EVA: channels, super-app, and Martijn's experiments. Two of them suffer from a lack of retransmitting `WriteRequest` and assume that `WriteRequest` should be retransmitted in case of failure.

So, my line of reasoning was wrong.

**Comment by drew2a** [comment]
See also this finding from @kozlovsky https://github.com/Tribler/tribler/issues/6845

**Comment by rmadhwal** [comment]
That seems like a bug indeed, but isn't the issue that I observed in practice in the kotlin implementation, rather than packets arriving out of order, during failures they don't seem to arrive at all.

But a potential patch to the kotlin library should indeed address that too, thanks for pointing it out!

**Comment by rmadhwal** [comment]
Beyond the above, we also seem to notice an OutOfMemory error when trying to send a large file over EVA, is that also something you guys experienced in Python ? I suspect a possible memory leak somewhere since the files should be broken into blocks so it's hard to guess why this occurs otherwise (Could also be specfic to the Kotlin implementation)

**Comment by devos50** [comment]
I haven't seen this behaviour in my experiments where I'm using EVA to regularly send ~300KB items between peers (using the Python implementation). @drew2a maybe you have seen something related to this?

**Comment by drew2a** [comment]
@rmadhwal @devos50 
no, we have never seen the `OutOfMemory` error in Python. 
I have a regular test for 1Mb transfer [here](https://github.com/Tribler/tribler/blob/a54200cc29aba05b763e3277c7cb344accc8a719/src/tribler/core/components/ipv8/tests/test_eva_protocol.py#L147-L157) maybe some modification of this test will be helpful to investigate this memory leak problem.

```python
    async def test_one_megabyte_transfer(self):
        # In this test we send `1Mb` transfer from Alice to Bob.
        data_size = 1024 * 1024
        data = os.urandom(1), os.urandom(data_size), random.randrange(0, 256)

        await self.alice.eva.send_binary(self.bob.my_peer, *data)

        assert self.bob.most_recent_received_data == data

        await self.alice.data_has_been_sent.wait()
        assert self.alice.most_recent_sent_data == data
```

BTW how big are the files that you trying to send over `EVA`?

@rmadhwal if you give me a link to the Kotlin implementation of EVA, I (probably) can check it on potential problems.

**Comment by drew2a** [comment]
@rmadhwal did you get this error by running an emulator or a real device?

My guess this error is not related to protocol but to the environment.

@devos50 @rmadhwal  what do you think guys, maybe it is better to use LibTorrent to transfer big files instead of EVA? EVA is relatively slow and has not been designed for this type of transfer.

**Comment by rmadhwal** [comment]
@drew2a I experienced the same behaviour on both.

We're using EVA as fallback because torrenting fails for us across networks so EVA seems to be the only option for now. Though if we want to restrict EVA to smaller size transfers, our team could look at splitting the file up and sending it piecewise through EVA which would be easier on the memory.

**Comment by devos50** [comment]
@drew2a I also tried libtorrent in my experiments but that was quite complicated to setup and requires additional information to be gossiped around (e.g., the port the libtorrent session is listening on). The main advantage is that libtorrent is able to parallelise the download of particular pieces, and has congestion control (either provided by TCP or uTP). For the moment being, EVA seems to do the job for the things I want to achieve. Specifically, I assume that the network progresses in synchronous rounds so the absolute completion time of an EVA transfer doesn't matter for protocol correctness. But I might drop that assumption later and I still require that the EVA protocol is robust.

Having said that, the tradeoff here seems to be programming convenience vs. transfer speed. On the one hand, When speed really matters, I would go for libtorrent but that requires (much) more time to properly setup and manage. Additionally, a libtorrent download also has some ramp-up time. So when transferring many small files in a small period, EVA might be faster. On the other hand, I would recommend EVA if one needs to send small files quickly, quickly need to prototype a transfer and/or when speed is an less important factor.

**Comment by drew2a** [comment]
@rmadhwal 

> Though if we want to restrict EVA to smaller size transfers, our team could look at splitting the file up and sending it piecewise through EVA which would be easier on the memory.

Yes. Another option is to use a file instead of memory for storing transferred data. It looks like an easy change to do.
